### ISSUE TITLE
Sweep: Transformers Integration

### ISSUE THREAD

**Issue by mosshammer** [mosshammer/sentence-transformers#2]
### Details

This repository currently relies on a fit method with a custom training loop. Despite working very well, user needs have outgrown that method. I believe that the transformers Trainer would be a very future-proof approach of improving the training in Sentence Transformers.

Ideally, I would like to introduce a new Trainer that implements a superclass of all behaviour supported by the current fit method. Then, I can fully refactor the fit method to use the new Trainer below the hood. This should prevent any backwards compatibility issues & allow training in the same way as before, although a new way (via the Trainer itself) will now also be supported. Training via the Trainer should allow for more opportunities to tune the training arguments.

## Pro's
This feature request should allow for better future-proofing by adopting powerful functionality from transformers whose interface they would not quickly change.
Sentence Transformers gains all benefits of the transformers Trainer. Some notable ones:
MultiGPU support.
Loss logging via W&B or Tensorboard & the terminal.
bf16 training.

## Cons
The Trainer is not explicitly created for embedding tasks, so we have to override certain "behind-the-scenes" code, i.e. code that could be changed from version to version & break the Sentence Transformers training.
The Trainer likes to work with Dataset instances, which may require sunsetting the InputExample class in favor of preprocessing data with datasets.
It may be challenging to avoid breaking changes, e.g. training with both positive pairs & text-positive-negative triplets might be possible currently (Not 100% confident), and we'd like to keep that behaviour possible if it currently is, but that will be difficult due to Datasets and their fixed columns.
It may be challenging to build the Trainer into a complete superset of all fit functionality, e.g. fit allows multi-task learning with multiple datasets and loss functions.
